_posts/2020-07-08-event-driven-functions-with-openfaas-and-nats.md

```diff
@@ -0,0 +1,146 @@
+* [VMware vCenter connector](https://github.com/openfaas-incubator/openfaas-vcenter-connector) an OpenFaaS event-connector built to consume events from vCenter and to trigger functions.
+
+There are several other connectors which allows you to trigger OpenFaaS functions based on events, are written and managed as a third party project. Please refer to this [link](https://docs.openfaas.com/reference/triggers/) for full list.
+
+## NATS
+
+[NATS](https://nats.io) is a simple, secure and high performance open source messaging system for cloud native applications, IoT messaging, and microservices architectures. In this blog post, I will be using NATS publish-subscribe concept where publishers publishes a message on a topic/subject and subscribers consumes that message by subscribing to that topic/subject.
+
+![NATS Publish Subscribe](/images/2020-openfaas-nats/nats-publish-subscribe.png "https://docs.nats.io/nats-concepts/pubsub")
+
+## Pre-requisite Installation
+
+* [k3d](https://github.com/rancher/k3d) to create a [k3s](https://github.com/rancher/k3s) cluster. k3s is lightweight distribution of kubernetes from Rancher Labs.
+* `kubectl` to manage kubernetes cluster.
+* `helm` to install `nats-connector` using helm chart
+* [arkade](https://github.com/alexellis/arkade) to install OpenFaaS. `arkade` is simple CLI tool to install helm charts and apps to your cluster in one command.
+* [faas-cli](https://github.com/openfaas/faas-cli) CLI tool to manage OpenFaaS functions.
+
+> Please make sure you have these tools installed before you proceed next.
+
+## Install Components
+
+#### Create a kubernetes cluster
+
+`k3d` is a tool to run `k3s` in docker. Create a single node `k3s` cluster.
+
+```
+k3d create
+```
+
+Follow instructions returned in the output of this command to configure KUBECONFIG to access the cluster using kubectl.
+
+#### Install OpenFaaS 
+Install OpenFaaS using `arkade`
+
+```
+arkade install openfaas --set basic_auth=true --set functionNamespace=openfaas-fn
+```
+
+Follow instructions printed to obtain the gateway password and login with `faas-cli`
+
+#### Install NATS connector using helm chart
+`nats-connector` is connector which invokes OpenFaaS function when a message is sent to a NATS topic.
+
+```
+helm repo add nats-connector  https://openfaas.github.io/faas-netes
+
+helm install nats-connector openfaas/nats-connector --namespace=openfaas
+
+```
+
+> NATS comes with OpenFaaS installation, so we are not setting up NATS here
+
+#### Login to OpenFaaS gateway using CLI
+
+Port forward the gateway service to access it. 
+
+```
+kubectl port-forward -n openfaas svc/gateway 8080:8080 &
+```
+
+Get the password
+
+```
+PASSWORD=$(kubectl get secret -n openfaas basic-auth -o jsonpath="{.data.basic-auth-password}" | base64 --decode; echo)
+```
+
+Login to the gateway
+
+```
+echo -n $PASSWORD | faas-cli login --username admin --password-stdin
+```
+
+
+## Deploy functions
```

Comment:
I wonder if we can come up with a more compelling use case for this pub-sub.  Perhaps we can have a function (or several functions) that emit "audit events" to an `audit` topic and a second topic that sends "alerts" based on the event?  This case is much easier to write in isolation and demo, so i 

if you do that, I think this section should be kept, but it would be better to call it "Verify the installation"

Reply:
@alexellis wanted to have a blog post showing how can nats connector be used to invoke functions for the new users. More like how to blog.

To demonstrate a pub-sub or event driven use case, we can have a separate blog where we can focus more on the usecase and implementation patterns